### PR TITLE
Geomap: Custom markers using ResourcePicker with standard marker fallbacks

### PR DESCRIPTION
**Pull request by nikki-kiga**
**What this PR does / why we need it**:
- Uses ResourceDimensionEditor and Picker to select custom icons
- Uses regular shape ol styles for some of the standard markers

**Special notes for your reviewer**:
- Relocated some of the geomap types
- Unsure whether to remove the Marker Shape select option or to leave as well

geo-icons
star-svg-marker-map

### CODE REVIEW

**Comment by ryantxu** [comment]
Nice!

This adds an "Icon path" attribute, but I think it should replace the "Marker shape" attribute.  We just want one setting rather than two.  BUT, when circle/cross/square/etc are selected, use a `RegularShape` implementation rather than the loaded svg.

In another PR we can look at removing/hiding the 'field' option in the editor

File: public/app/features/dimensions/types.ts [hunk 1]
```diff
@@ -66,6 +66,7 @@ export interface ColorDimensionConfig extends BaseDimensionConfig<string> {}
 /** Places that use the value */
 export interface ResourceDimensionOptions {
   resourceType: 'icon' | 'image';
+  folderIndex: number;
```
Comment (ryantxu) [inline public/app/features/dimensions/types.ts:69]: rather than index... maybe have the folder name?  

Also `showSourceRadio` should be here not in the saved model

**Review by ryantxu — COMMENTED** [review]
We are getting close!

Rather than call this "Icon path"
image

Can it just replace the existing "symbol" place in the UI?

To support migrations, we will need to add a migration handler like:
https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/table/module.tsx#L16
and
https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/table/migrations.ts#L18

this should change the existing "triangle" config to the common path

File: public/app/features/dimensions/editors/ResourceDimensionEditor.tsx [hunk 2]
```diff
@@ -57,22 +57,33 @@ export const ResourceDimensionEditor: FC<
     setOpen(true);
   }, []);
 
+  console.log('source', value?.showSourceRadio);
+  console.log('mode', value?.mode);
```
Comment (hugohaggmark) [inline public/app/features/dimensions/editors/ResourceDimensionEditor.tsx:61]: nit: please remove.

File: public/app/features/dimensions/editors/ResourcePicker.tsx [hunk 3]
```diff
@@ -33,12 +34,17 @@ interface ResourceItem {
...
 
 export function ResourcePicker(props: Props) {
-  const { value, onChange, mediaType } = props;
-  const folders = (mediaType === 'icon' ? ['img/icons/unicons', 'img/icons/iot'] : ['img/bg']).map((v) => ({
-    label: v,
-    value: v,
-  }));
-  const folderOfCurrentValue = value ? folders.filter((folder) => value.indexOf(folder.value) > -1)[0] : folders[0];
+  const { value, onChange, mediaType, folderIndex } = props;
+  const folders = (mediaType === 'icon' ? ['img/icons/unicons', 'img/icons/iot', 'img/icons/geo'] : ['img/bg']).map(
```
Comment (hugohaggmark) [inline public/app/features/dimensions/editors/ResourcePicker.tsx:38]: nit: add these to a constant.ts file to make it easier to change moving forward?

File: public/app/plugins/panel/geomap/layers/basemaps/carto.ts [hunk 4]
```diff
@@ -52,7 +53,7 @@ export const carto: MapLayerRegistryItem<CartoConfig> = {
     },
   }),
 
-  registerOptionsUI: (builder) => {
+  registerOptionsUI: (builder: any) => {
```
Comment (hugohaggmark) [inline public/app/plugins/panel/geomap/layers/basemaps/carto.ts:56]: I don't think we need all these `anys` in this PR.

File: public/app/plugins/panel/geomap/layers/data/markersLayer.tsx [hunk 5]
```diff
@@ -88,16 +93,27 @@ export const markersLayer: MapLayerRegistryItem<MarkersConfig> = {
     if (config.showLegend) {
       legend = <ObservablePropsWrapper watch={legendProps} initialSubProps={{}} child={MarkersLegend} />;
     }
-    const shape = markerMakers.getIfExists(config.shape) ?? circleMarker;
+
+    let shape = markerMakers.getIfExists(config.shape) ?? circleMarker;
```
Comment (hugohaggmark) [inline public/app/plugins/panel/geomap/layers/data/markersLayer.tsx:97]: nit: we should try to avoid `let` if possible. Makes the code hard to follow and complex.

**Review by hugohaggmark — COMMENTED** [review]
I did a shallow code review and it looks good.

File: public/app/plugins/panel/geomap/types.ts [hunk 6]
```diff
@@ -37,6 +47,48 @@ export const defaultView: MapViewConfig = {
   zoom: 1,
 };
 
+/**
+ * @alpha
```
Comment (mckn) [inline public/app/plugins/panel/geomap/types.ts:51]: The documentation code comments can be removed since these are now living in core grafana.

**Review by mckn — COMMENTED** [review]
Overall I think this looks good! Just make sure that the types removed from the `@grafana/data` is safe to delete.

**Comment by ryantxu** [comment]
Another minor nit... can the folder name be "map" or "marker" rather than "geo"
image

**Review by ryantxu — APPROVED** [review]
looks good!  thanks @nikki-kiga

**Comment by ryantxu** [comment]
@hugohaggmark see https://github.com/grafana/grafana/pull/40237
